src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java

```diff
@@ -0,0 +1,122 @@
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.ca.plugins;
+
+import com.mysql.cj.log.Log;
+
+public class MonitorConnectionContext {
+
+  private boolean isConnectionValid;
+  private int failureCount;
+  private int failureDetectionTimeMillis;
+  private int failureDetectionIntervalMillis;
+  private int failureDetectionCount;
+  private final String node;
+  private final Log log;
+
+  public MonitorConnectionContext(
+      String node,
+      Log log,
+      int failureDetectionTimeMillis,
+      int failureDetectionIntervalMillis,
+      int failureDetectionCount) {
+    this.node = node;
+    this.log = log;
+    this.failureDetectionTimeMillis = failureDetectionTimeMillis;
+    this.failureDetectionIntervalMillis = failureDetectionIntervalMillis;
+    this.failureDetectionCount = failureDetectionCount;
+  }
+
+  public int getFailureDetectionTimeMillis() {
+    return failureDetectionTimeMillis;
+  }
+
+  public void setFailureDetectionTimeMillis(int failureDetectionTimeMillis) {
+    this.failureDetectionTimeMillis = failureDetectionTimeMillis;
+  }
+
+  public int getFailureDetectionIntervalMillis() {
+    return failureDetectionIntervalMillis;
+  }
+
+  public void setFailureDetectionIntervalMillis(int failureDetectionIntervalMillis) {
+    this.failureDetectionIntervalMillis = failureDetectionIntervalMillis;
+  }
+
+  public int getFailureDetectionCount() {
+    return failureDetectionCount;
+  }
+
+  public void setFailureDetectionCount(int failureDetectionCount) {
+    this.failureDetectionCount = failureDetectionCount;
+  }
+
+  public int getFailureCount() {
+    return failureCount;
+  }
+
+  public void setFailureCount(int failureCount) {
+    this.failureCount = failureCount;
+  }
+
+  public void incrementFailureCount() {
+    this.failureCount++;
+  }
+
+  public boolean isConnectionValid() {
+    return isConnectionValid;
+  }
+
+  public void setConnectionValid(boolean connectionValid) {
+    isConnectionValid = connectionValid;
+  }
+
+  public boolean isNodeUnhealthy() {
```

Comment:
I think all this logic should be moved to setConnectionValid().
isNodeUnhealthy() is supposed to return just a pre-calculated flag like isConnectionValid()